src/nunavut/lang/c/support/serialization.j2

```diff
@@ -161,26 +177,48 @@ static inline void nunavutCopyBits(void* const dst,
-        // offsets be under 8 bits. Fewer constraints reduce the chance of API misuse.
-        const uint8_t* const psrc = (src_offset_bits / 8U) + (const uint8_t*) src;  // NOSONAR NOLINT
-        uint8_t*       const pdst = (dst_offset_bits / 8U) +       (uint8_t*) dst;  // NOSONAR NOLINT
-        (void) memmove(pdst, psrc, length_bytes);
-        const uint8_t length_mod = (uint8_t)(length_bits % 8U);
-        if (0U != length_mod)  // If the length is unaligned, the last byte requires special treatment.
+        const char* last_src;
+        char*       last_dst;
+        unsigned    mask, length_mod;
+        const {{ typename_unsigned_length }} length_chars = ({{ typename_unsigned_length }})(length_bits >> CHAR_SHIFT);
+        const char* const psrc    = (src_offset_bits >> CHAR_SHIFT) + (const char*)src;
+        char*       const pdst    = (dst_offset_bits >> CHAR_SHIFT) + (      char*)dst;
+        (void)memmove(pdst, psrc, length_chars);
+        length_mod = length_bits & CHAR_SHIFT_MASK;
+        if (length_mod == 0)
+            return;
+        last_src  = psrc + length_chars;
+        last_dst  = pdst + length_chars;
+        mask      = (unsigned)((1U << length_mod) - 1U);
+        *last_dst = (*last_dst & (unsigned)(~mask)) | (*last_src & mask);
+        return;
+    }
+    #if ((-1) >> 1) < 0  // Right shift is arithmetic (the most common case). So, the fast algorithm is allowed
+    {
+        const char* const psrc = (const char*)src;
+        char*       const pdst = (      char*)dst;
+        {{ typename_unsigned_bit_length }} src_off = src_offset_bits;
+        {{ typename_unsigned_bit_length }} dst_off = dst_offset_bits;
+        const {{ typename_unsigned_bit_length }} last_bit = src_off + length_bits;
+        while (last_bit > src_off)
         {
-            // Intentional violation of MISRA: Pointer arithmetics. It is unavoidable in this context.
-            const uint8_t* const last_src = psrc + length_bytes;  // NOLINT NOSONAR
-            uint8_t* const last_dst       = pdst + length_bytes;  // NOLINT NOSONAR
-            {{ assert('length_mod < 8U') }}
-            const uint8_t mask = (uint8_t)((1U << length_mod) - 1U);
-            *last_dst = (*last_dst & (uint8_t)~mask) | (*last_src & mask);
+            const unsigned src_mod = src_off & CHAR_SHIFT_MASK;
+            const unsigned dst_mod = dst_off & CHAR_SHIFT_MASK;
+            const unsigned max_mod = src_mod > dst_mod ? src_mod : dst_mod;
+            const unsigned size    = nunavutChooseMin(CHAR_BIT - max_mod, last_bit - src_off);
+            const unsigned mask    = (((1U << size) - 1) << dst_mod) & UCHAR_MAX;
+            const unsigned in      = ((unsigned)(psrc[src_off >> CHAR_SHIFT] >> src_mod) << dst_mod) & UCHAR_MAX;
+            const unsigned a       = pdst[dst_off >> CHAR_SHIFT] & (~mask);
+            const unsigned b       = in & mask;
+            pdst[dst_off >> CHAR_SHIFT] = a | b;
```

Comment:
this assignment loses precision. Again, a lot of compiler warnings need to be cleaned up.

Reply:
> Again, a lot of compiler warnings need to be cleaned up.

I use the default mingw settings. They're pretty strict. There are no warnings. I we'll have to  investigate

Reply:
I'm going to push an update to Nunavut HEAD in a bit that uses the latest dev container which has a very new version of clang I was building against. You'll be able to find these warning by using that. Regardless, once you start looking you can see a lot of signed/unsigned conversions going on and possible precision loss.

Reply:
See #297